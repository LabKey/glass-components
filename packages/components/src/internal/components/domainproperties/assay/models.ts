/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { List, Record as ImmutableRecord } from 'immutable';

import { getServerContext, Utils } from '@labkey/api';

import { DomainDesign, FieldErrors } from '../models';
import { AppURL } from '../../../url/AppURL';
import { getAppHomeFolderPath } from '../../../app/utils';
import { Container } from '../../base/models/Container';

// See ExpProtocol.Status in 'platform' repository.
export enum Status {
    Active = 'Active',
    Archived = 'Archived',
}

export interface ProtocolTransformScript {
    runOnEdit: boolean;
    runOnImport: boolean;
    scriptPath: string;
}

export class AssayProtocolModel extends ImmutableRecord({
    allowBackgroundUpload: false,
    allowEditableResults: false,
    allowQCStates: false,
    allowTransformationScript: false,
    allowPlateMetadata: false,
    autoCopyTargetContainer: undefined,
    autoCopyTargetContainerId: undefined,
    autoLinkCategory: undefined,
    availableDetectionMethods: undefined,
    availableMetadataInputFormats: undefined,
    availablePlateTemplates: undefined,
    backgroundUpload: false,
    description: undefined,
    domains: undefined,
    editableResults: false,
    editableRuns: false,
    exception: undefined,
    metadataInputFormatHelp: undefined,
    moduleTransformScripts: undefined,
    name: undefined,
    plateMetadata: undefined,
    protocolId: undefined,
    protocolParameters: undefined,
    protocolTransformScripts: undefined,
    providerName: undefined,
    saveScriptFiles: false,
    selectedDetectionMethod: undefined,
    selectedMetadataInputFormat: undefined,
    selectedPlateTemplate: undefined,
    status: Status.Active,
    qcEnabled: undefined,
    excludedContainerIds: undefined,
}) {
    declare allowBackgroundUpload: boolean;
    declare allowEditableResults: boolean;
    declare allowQCStates: boolean;
    declare allowTransformationScript: boolean;
    declare allowPlateMetadata: boolean;
    declare autoCopyTargetContainer: {};
    declare autoCopyTargetContainerId: string;
    declare autoLinkCategory: string;
    declare availableDetectionMethods: [];
    declare availableMetadataInputFormats: {};
    declare availablePlateTemplates: [];
    declare backgroundUpload: boolean;
    declare description: string;
    declare domains: List<DomainDesign>;
    declare editableResults: boolean;
    declare editableRuns: boolean;
    declare exception: string;
    declare metadataInputFormatHelp: any;
    declare moduleTransformScripts: List<string>;
    declare name: string;
    declare plateMetadata: boolean;
    declare protocolId: number;
    declare protocolParameters: any;
    declare protocolTransformScripts: List<ProtocolTransformScript>;
    declare providerName: string;
    declare saveScriptFiles: boolean;
    declare selectedDetectionMethod: string;
    declare selectedMetadataInputFormat: string;
    declare selectedPlateTemplate: string;
    declare status: Status;
    declare qcEnabled: boolean;
    declare excludedContainerIds: string[];

    static create(raw: any): AssayProtocolModel {
        let domains = raw.domains || List<DomainDesign>();
        if (raw.domains && Utils.isArray(raw.domains)) {
            domains = List<DomainDesign>(
                raw.domains.map(domain => {
                    return DomainDesign.create(domain);
                })
            );
        }

        if (raw.protocolTransformScripts && Utils.isArray(raw.protocolTransformScripts)) {
            raw.protocolTransformScripts = List(raw.protocolTransformScripts);
        }
        if (raw.moduleTransformScripts && Utils.isArray(raw.moduleTransformScripts)) {
            raw.moduleTransformScripts = List<string>(raw.moduleTransformScripts);
        }

        // if this is not an existing assay, clear the name property so the user must set it
        const name = !raw.protocolId ? undefined : raw.name;

        // Issue 38685: for new assays, pre-select some required assay properties
        const model = new AssayProtocolModel({ ...raw, name, domains });
        if (model.isNew()) {
            if (model.allowDetectionMethodSelection() && model.availableDetectionMethods.length > 0) {
                raw.selectedDetectionMethod = List<string>(model.availableDetectionMethods).get(0);
            }
            if (model.allowPlateTemplateSelection() && model.availablePlateTemplates.length > 0) {
                raw.selectedPlateTemplate = List<string>(model.availablePlateTemplates).get(0);
            }
        }

        return new AssayProtocolModel({ ...raw, name, domains });
    }

    static serialize(model: AssayProtocolModel): any {
        // need to serialize the DomainDesign objects to remove the unrecognized fields
        const domains = model.domains.map(domain => {
            return DomainDesign.serialize(domain);
        });

        const json = model.merge({ domains }).toJS();

        // only need to serialize the id and not the autoCopyTargetContainer object
        delete json.autoCopyTargetContainer;
        delete json.exception;
        return json;
    }

    isActive(): boolean {
        return this.status === Status.Active;
    }

    getDomainByNameSuffix(name: string): DomainDesign {
        if (this.domains.size > 0) {
            return this.domains.find(domain => {
                return domain.isNameSuffixMatch(name);
            });
        }
    }

    get container(): string {
        const container = new Container(getServerContext().container);
        const domainContainerId = this.domainContainerId;

        return this.isNew()
            ? getAppHomeFolderPath(container)
            : domainContainerId === container.id
              ? container.path
              : domainContainerId;
    }

    get domainContainerId(): string {
        return this.getIn(['domains', 0, 'container']);
    }

    isNew(): boolean {
        return !this.protocolId;
    }

    allowPlateTemplateSelection(): boolean {
        return this.availablePlateTemplates && Utils.isArray(this.availablePlateTemplates);
    }

    allowDetectionMethodSelection(): boolean {
        return this.availableDetectionMethods && Utils.isArray(this.availableDetectionMethods);
    }

    allowMetadataInputFormatSelection(): boolean {
        return (
            this.availableMetadataInputFormats &&
            Utils.isObject(this.availableMetadataInputFormats) &&
            !Utils.isEmptyObj(this.availableMetadataInputFormats)
        );
    }

    validateTransformScripts(): string {
        if (this.protocolTransformScripts === undefined || this.protocolTransformScripts.size === 0) {
            return undefined;
        }

        // make sure we don't have any script inputs that are empty strings
        const hasEmptyScript = this.protocolTransformScripts.some(
            config => config?.scriptPath === undefined || config.scriptPath === null || config.scriptPath.length === 0
        );
        if (hasEmptyScript) {
            return 'Missing required transform script path.';
        }
    }

    getFirstDomainFieldError(): FieldErrors {
        const firstErrantDomain = this.domains.find(domain => domain.hasInvalidFields());
        return firstErrantDomain !== undefined ? firstErrantDomain.getInvalidFields().first().getErrors() : undefined;
    }

    isValid(): boolean {
        return this.hasValidProperties() && this.getFirstDomainFieldError() === undefined;
    }

    isGPAT(): boolean {
        return this.providerName.toLowerCase() === 'general';
    }

    hasValidProperties(): boolean {
        return (
            this.name !== undefined &&
            this.name !== null &&
            this.name.trim().length > 0 &&
            (!this.allowMetadataInputFormatSelection() || Utils.isString(this.selectedMetadataInputFormat)) &&
            (!this.allowDetectionMethodSelection() || Utils.isString(this.selectedDetectionMethod)) &&
            (!this.allowPlateTemplateSelection() || Utils.isString(this.selectedPlateTemplate)) &&
            this.validateTransformScripts() === undefined
        );
    }

    get hasBatchFields(): boolean {
        return this.getDomainByNameSuffix('Batch')?.fields.size > 0;
    }

    getUrl(): AppURL {
        return AppURL.create('assays', this.providerName, this.name);
    }
}
