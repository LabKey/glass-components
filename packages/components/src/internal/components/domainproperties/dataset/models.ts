/*
 * Copyright (c) 2020 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Record } from 'immutable';

import { immerable, produce } from 'immer';

import { DomainDesign, DomainField } from '../models';

import { DOMAIN_FIELD_FULLY_LOCKED } from '../constants';

import { allowAsManagedField } from './utils';

export interface DatasetAdvancedSettingsForm {
    cohortId?: number;
    datasetId?: number;
    showByDefault?: boolean;
    tag?: string;
    visitDatePropertyName?: string;
}

export interface IDatasetModel {
    category?: string;
    cohortId?: number;
    dataSharing?: string;
    datasetId?: number;
    definitionIsShared?: boolean;
    demographicData: boolean;
    description?: string;
    domain: DomainDesign;
    domainId: number;
    entityId?: string;
    exception: string;
    keyPropertyManaged: boolean;
    keyPropertyName?: string;
    tag?: string;
    visitDatePropertyName?: string;
    showByDefault: boolean;
    sourceName?: string;
    sourceType?: string;
    name: string;
    sourceUrl?: string;
    label?: string;
    useTimeKeyField?: boolean;
}

export class DatasetModel implements IDatasetModel {
    [immerable] = true;

    readonly domain: DomainDesign;
    readonly domainId: number;
    readonly exception: string;
    readonly datasetId?: number;
    readonly entityId?: string;
    readonly name: string;
    readonly category?: string;
    readonly visitDatePropertyName?: string;
    readonly keyPropertyName?: string;
    readonly keyPropertyManaged: boolean;
    readonly demographicData: boolean;
    readonly label?: string;
    readonly cohortId?: number;
    readonly tag?: string;
    readonly showByDefault: boolean;
    readonly description?: string;
    readonly dataSharing?: string;
    readonly definitionIsShared?: boolean;
    readonly sourceName?: string;
    readonly sourceUrl?: string;
    readonly sourceType?: string;
    readonly useTimeKeyField?: boolean;

    constructor(datasetModel: IDatasetModel) {
        Object.assign(this, datasetModel);
    }

    static create(newDataset = null, raw?: any): DatasetModel {
        if (newDataset) {
            const domain = DomainDesign.create(undefined);
            return new DatasetModel({ ...newDataset, domain });
        } else {
            const domain = DomainDesign.create(raw.domainDesign);
            let model = new DatasetModel({ ...raw.options, domain });

            // if the dataset is from a linked source, disable/lock the fields
            if (model.isFromLinkedSource()) {
                const newDomain = domain.merge({
                    fields: domain.fields
                        .map((field: DomainField) => {
                            return field.set('lockType', DOMAIN_FIELD_FULLY_LOCKED);
                        })
                        .toList(),
                }) as DomainDesign;

                model = produce<DatasetModel>(model, draft => {
                    draft.domain = newDomain;
                });
            }

            return model;
        }
    }

    hasValidProperties(): boolean {
        const isValidName = this.name !== undefined && this.name !== null && this.name.trim().length > 0;
        const isValidLabel =
            this.isNew() || (this.label !== undefined && this.label !== null && this.label.trim().length > 0);

        return isValidName && isValidLabel && this.hasValidAdditionalKey();
    }

    hasValidAdditionalKey(): boolean {
        let isValidKeySetting = true;
        if (this.getDataRowSetting() === 2) {
            isValidKeySetting =
                (this.keyPropertyName !== undefined && this.keyPropertyName !== '') || this.useTimeKeyField;
        }

        return isValidKeySetting;
    }

    isNew(): boolean {
        return !this.entityId;
    }

    getDataRowSetting(): number {
        const noKeyPropName = this.keyPropertyName === undefined || this.keyPropertyName === null;

        // participant id
        if (noKeyPropName && this.demographicData) {
            return 0;
        }
        // participant id and timepoint
        else if (noKeyPropName && !this.useTimeKeyField) {
            return 1;
        }

        // participant id, timepoint and additional key field
        return 2;
    }

    validManagedKeyField(editedName?: string): boolean {
        if (this.keyPropertyName || editedName) {
            const domainFields = this.domain.fields;

            const allowedFieldTypes = domainFields
                .filter(field => allowAsManagedField(field))
                .map(field => {
                    return field.name;
                })
                .toList();

            return allowedFieldTypes.contains(this.keyPropertyName) || allowedFieldTypes.contains(editedName);
        } else {
            return false;
        }
    }

    getDomainKind(timepointType: string): string {
        return timepointType === 'VISIT' ? 'StudyDatasetVisit' : 'StudyDatasetDate';
    }

    getOptions(): Record<string, any> {
        return produce<IDatasetModel>(this, draft => {
            delete draft.exception;
            delete draft.domain;
        });
    }

    isValid(): boolean {
        return this.hasValidProperties() && !this.domain.hasInvalidFields();
    }

    isFromLinkedSource(): boolean {
        return this.sourceName !== undefined && this.sourceName !== null;
    }
}
