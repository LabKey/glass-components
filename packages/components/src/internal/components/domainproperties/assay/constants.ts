export const PROPERTIES_HEADER_ID = 'assay-properties-hdr';
export const FORM_ID_PREFIX = 'assay-design-';

export const FORM_IDS = {
    ASSAY_NAME: FORM_ID_PREFIX + 'name',
    ASSAY_DESCRIPTION: FORM_ID_PREFIX + 'description',
    AUTO_LINK_TARGET: FORM_ID_PREFIX + 'autoCopyTargetContainerId',
    AUTO_LINK_CATEGORY: FORM_ID_PREFIX + 'autoLinkCategory',
    BACKGROUND_UPLOAD: FORM_ID_PREFIX + 'backgroundUpload',
    DETECTION_METHOD: FORM_ID_PREFIX + 'selectedDetectionMethod',
    EDITABLE_RUNS: FORM_ID_PREFIX + 'editableRuns',
    EDITABLE_RESULTS: FORM_ID_PREFIX + 'editableResults',
    METADATA_INPUT_FORMAT: FORM_ID_PREFIX + 'selectedMetadataInputFormat',
    PLATE_TEMPLATE: FORM_ID_PREFIX + 'selectedPlateTemplate',
    PROTOCOL_TRANSFORM_SCRIPTS: FORM_ID_PREFIX + 'protocolTransformScripts',
    QC_ENABLED: FORM_ID_PREFIX + 'qcEnabled',
    SAVE_SCRIPT_FILES: FORM_ID_PREFIX + 'saveScriptFiles',
    PLATE_METADATA: FORM_ID_PREFIX + 'plateMetadata',
    STATUS: FORM_ID_PREFIX + 'status',
};

export const BOOLEAN_FIELDS = [
    FORM_IDS.BACKGROUND_UPLOAD,
    FORM_IDS.EDITABLE_RUNS,
    FORM_IDS.EDITABLE_RESULTS,
    FORM_IDS.QC_ENABLED,
    FORM_IDS.SAVE_SCRIPT_FILES,
    FORM_IDS.PLATE_METADATA,
    FORM_IDS.STATUS,
];
