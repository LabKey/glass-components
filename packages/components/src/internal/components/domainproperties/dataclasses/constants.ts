export const DATACLASS_DOMAIN_SYSTEM_FIELDS = [
    {
        Name: 'Name',
        Label: 'Name',
        'Data Type': 'Text',
        Required: false,
        Description: 'Contains a short name for this data object.',
    },
    {
        Name: 'Description',
        Label: 'Description',
        'Data Type': 'Text',
        Required: false,
        Description: 'Contains a description for this data object',
    },
];
