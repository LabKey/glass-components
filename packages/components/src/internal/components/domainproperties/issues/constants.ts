export const ISSUES_LIST_DEF_SINGULAR_PLURAL_TIP =
    'The Singular item name and Plural item name fields control the display ' +
    'name for an "issue" across the Issues module. For example, you might instead refer to issues as "Bugs", "Tickets", or "Opportunities" depending on the context.';

export const ISSUES_LIST_DEF_SORT_DIRECTION_TIP =
    'By default, comments on an issue are shown in the order they are added, oldest first. Change the Comment Sort Direction to newest first if you prefer.';
export const ISSUES_LIST_GROUP_ASSIGN_TIP =
    "Select a specific group from the dropdown, both site and project level groups are available. If left blank, the 'Assigned To' field will display all users who are a) assigned the role of Editor or higher in the folder, and b) are members of at least one project-level security group.";
export const ISSUES_LIST_USER_ASSIGN_TIP =
    'In some workflows it is useful to have a default user to whom all issues are assigned, such as for initial triage and balancing assignments across a group. If left blank, no user will be assigned by default.';
export const ISSUES_LIST_RELATED_FOLDER_TIP =
    'A related issue can be created in any issue list that the user can insert into. This setting controls the default folder selection in the related issue dialog box.';
export const ISSUES_LIST_RESTRICTED_TRACKER_TIP =
    'A restricted issue list will limit viewing and updating of all issues in the list. When the checkbox is selected only users who are either assigned to the issue or on the notify list ' +
    'will have access to an issue. An additional group can optionally be configured to provide access to any members of that group.';
export const ISSUES_LIST_RESTRICTED_GROUP_TIP =
    'An optional project or site group that can give members of the group access to the issues in a restricted issue list.';
