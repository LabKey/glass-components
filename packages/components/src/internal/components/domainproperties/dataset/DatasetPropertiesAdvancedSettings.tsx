import React, { ReactNode } from 'react';

import { DomainDesignerCheckbox } from '../DomainDesignerCheckbox';

import { Modal } from '../../../Modal';
import { getSubmitButtonClass } from '../../../app/utils';

import { DATASET_PROPERTIES_TOPIC, HelpLink } from '../../../util/helpLinks';

import { SectionHeading } from '../SectionHeading';

import { DomainFieldLabel } from '../DomainFieldLabel';

import { SelectInput, SelectInputOption } from '../../forms/input/SelectInput';

import { LabelHelpTip } from '../../base/LabelHelpTip';

import { DatasetAdvancedSettingsForm, DatasetModel } from './models';
import { fetchCohorts, getHelpTip, getVisitDateColumns } from './actions';
import { StudyProperties } from './utils';
import { SHOW_IN_OVERVIEW } from './constants';

interface DatasetSettingsSelectProps {
    clearable?: boolean;
    disabled?: boolean;
    helpTip?: ReactNode;
    label: string;
    labelKey?: string;
    name: string;
    onSelectChange: (name, formValue, selected) => void;
    selectOptions: any;
    selectedValue?: any;
    valueKey?: string;
}

export class DatasetSettingsSelect extends React.PureComponent<DatasetSettingsSelectProps> {
    render() {
        const {
            name,
            label,
            helpTip,
            selectedValue,
            selectOptions,
            onSelectChange,
            labelKey,
            valueKey,
            disabled,
            clearable,
        } = this.props;

        return (
            <div className="row margin-top">
                <div className="col-xs-5">
                    <DomainFieldLabel label={label} helpTipBody={helpTip} />
                </div>

                <div className="col-xs-7">
                    <SelectInput
                        onChange={onSelectChange}
                        value={selectedValue}
                        options={selectOptions}
                        inputClass=""
                        containerClass=""
                        labelClass=""
                        name={name}
                        labelKey={labelKey}
                        valueKey={valueKey}
                        disabled={disabled}
                        clearable={clearable}
                    />
                </div>
            </div>
        );
    }
}

interface DatasetSettingsInputProps {
    disabled: boolean;
    helpTip: ReactNode;
    label: string;
    name: string;
    onValueChange: (evt: any) => any;
    placeholder?: string;
    required: boolean;
    showInAdvancedSettings: boolean;
    value?: any;
}

export class DatasetSettingsInput extends React.PureComponent<DatasetSettingsInputProps> {
    render() {
        const { name, label, helpTip, value, placeholder, onValueChange, disabled, showInAdvancedSettings, required } =
            this.props;

        return (
            <div className="row margin-top">
                <div className="col-xs-4">
                    <DomainFieldLabel label={label} required={required} helpTipBody={helpTip} />
                </div>

                {showInAdvancedSettings && <div className="col-xs-1" />}

                <div className="col-xs-7">
                    <input
                        className="form-control"
                        id={name}
                        type="text"
                        placeholder={placeholder}
                        value={value}
                        onChange={onValueChange}
                        disabled={disabled}
                    />
                </div>

                {!showInAdvancedSettings && <div className="col-xs-1" />}
            </div>
        );
    }
}

interface AdvancedSettingsProps {
    applyAdvancedProperties: (datasetAdvancedSettingsForm: DatasetAdvancedSettingsForm) => void;
    model: DatasetModel;
    studyProperties: StudyProperties;
    title: string;
    visitDatePropertyIndex?: number;
}

interface AdvancedSettingsState extends DatasetAdvancedSettingsForm {
    availableCohorts?: SelectInputOption | SelectInputOption[];
    dataSharing?: string;
    modalOpen?: boolean;
    visitDateColumns?: SelectInputOption | SelectInputOption[];
}

export class AdvancedSettings extends React.PureComponent<AdvancedSettingsProps, AdvancedSettingsState> {
    constructor(props) {
        super(props);
        const initialState = this.getInitialState();

        this.state = {
            modalOpen: false,
            ...initialState,
        } as AdvancedSettingsState;
    }

    componentDidMount(): void {
        fetchCohorts()
            .then(data => {
                this.setState({
                    availableCohorts: data.toArray(),
                });
            })
            .catch(error => {
                console.error('Failed to retrieve available cohorts.', error);
            });
    }

    getInitialState = () => {
        const model = this.props.model;

        return {
            datasetId: model.datasetId,
            tag: model.tag,
            showByDefault: model.showByDefault,
            cohortId: model.cohortId,
            dataSharing: model.dataSharing,
            visitDatePropertyName: this.getVisitDatePropertyName(),
        };
    };

    getVisitDatePropertyName(): string {
        const { model, visitDatePropertyIndex } = this.props;
        return visitDatePropertyIndex !== undefined
            ? model.domain.fields.get(visitDatePropertyIndex).name
            : model.visitDatePropertyName;
    }

    toggleModal = (isModalOpen: boolean): void => {
        this.setState({ modalOpen: isModalOpen });

        // If modal is re-opened, reset unsaved values
        if (isModalOpen) {
            this.setState(this.getInitialState());
        }
    };

    closeModal = () => {
        this.toggleModal(false);
    };

    onInputChange = e => {
        const id = e.target.id;
        let value = e.target.value;

        if (e.target.type === 'checkbox') {
            value = e.target.checked;
        }

        this.setState({ [id]: value });
    };

    onSelectChange = (id: string, value: any): void => {
        this.setState({ [id]: value });
    };

    getHelpTipElement(field: string): JSX.Element {
        return (<> {getHelpTip(field, this.props.studyProperties)} </>) as JSX.Element;
    }

    applyChanges = (): void => {
        const { datasetId, cohortId, visitDatePropertyName, showByDefault, tag, dataSharing } = this.state;

        const datasetAdvancedSettingsForm = {
            showByDefault,
            datasetId,
            cohortId,
            visitDatePropertyName,
            tag,
            dataSharing,
        };

        const { applyAdvancedProperties } = this.props;

        applyAdvancedProperties(datasetAdvancedSettingsForm as DatasetAdvancedSettingsForm);
        this.toggleModal(false);
    };

    render() {
        const { modalOpen, datasetId, cohortId, tag, showByDefault, dataSharing, availableCohorts } = this.state;
        const { model, title, studyProperties } = this.props;

        const showDataspace = model.definitionIsShared && model.getDataRowSetting() === 0;
        const showDataspaceCls = showDataspace ? 'dataset_data_row_element_show' : 'dataset_data_row_element_hide';
        const showInOverviewLabel = 'Show dataset in overview';
        const visitDateColumns = getVisitDateColumns(model.domain).toArray();
        const visitDateProperty = this.getVisitDatePropertyName();
        const footer = (
            <>
                <button
                    className="domain-adv-footer domain-adv-cancel-btn btn btn-default"
                    onClick={this.closeModal}
                    type="button"
                >
                    Cancel
                </button>

                <HelpLink topic={DATASET_PROPERTIES_TOPIC} className="domain-adv-footer domain-adv-link">
                    Get help with dataset settings
                </HelpLink>

                <button
                    className={`domain-adv-footer domain-adv-apply-btn btn btn-${getSubmitButtonClass()}`}
                    onClick={this.applyChanges}
                    type="button"
                >
                    Apply
                </button>
            </>
        );

        return (
            <>
                <button
                    className="domain-field-float-right btn btn-default"
                    onClick={() => this.toggleModal(true)}
                    type="button"
                >
                    {title}
                </button>

                {modalOpen && (
                    <Modal footer={footer} onCancel={this.closeModal} title="Advanced Dataset Settings">
                        <SectionHeading title="Miscellaneous Options" />

                        <div className="row margin-top">
                            <div className="col-xs-5">
                                {showInOverviewLabel}
                                <LabelHelpTip title={showInOverviewLabel}>{SHOW_IN_OVERVIEW}</LabelHelpTip>
                            </div>
                            <div className="col-xs-7">
                                <DomainDesignerCheckbox
                                    checked={showByDefault}
                                    onChange={this.onInputChange}
                                    id="showByDefault"
                                    className="domain-field-checkbox"
                                />
                            </div>
                        </div>

                        <DatasetSettingsInput
                            name="datasetId"
                            label="Dataset ID"
                            helpTip={this.getHelpTipElement('datasetId')}
                            value={datasetId}
                            placeholder="Auto Assign"
                            disabled={!model.isNew()}
                            onValueChange={this.onInputChange}
                            showInAdvancedSettings={true}
                            required={true}
                        />
                        {studyProperties.TimepointType === 'VISIT' && (
                            <DatasetSettingsSelect
                                name="visitDatePropertyName"
                                label="Visit Date Column"
                                helpTip={this.getHelpTipElement('visitDateColumn')}
                                selectOptions={visitDateColumns}
                                selectedValue={visitDateProperty}
                                onSelectChange={this.onSelectChange}
                            />
                        )}
                        {/** * TODO: Look into - Cohort- Query Select didn't work  ***/}
                        <DatasetSettingsSelect
                            name="cohortId"
                            label="Cohort Association"
                            helpTip={this.getHelpTipElement('cohort')}
                            selectOptions={availableCohorts}
                            selectedValue={cohortId}
                            onSelectChange={this.onSelectChange}
                        />

                        <DatasetSettingsInput
                            name="tag"
                            label="Tag"
                            helpTip={this.getHelpTipElement('tag')}
                            value={tag}
                            disabled={false}
                            onValueChange={this.onInputChange}
                            showInAdvancedSettings={true}
                            required={false}
                        />

                        {model.definitionIsShared && (
                            <>
                                <div className={showDataspaceCls}>
                                    <div className="margin-top">
                                        <SectionHeading title="Dataspace Project Options" />
                                    </div>

                                    <DatasetSettingsSelect
                                        name="dataSharing"
                                        label="Share demographic data"
                                        helpTip={this.getHelpTipElement('dataspace')}
                                        selectOptions={[
                                            { label: 'No', value: 'NONE' },
                                            { label: 'Share by ' + studyProperties.SubjectColumnName, value: 'PTID' },
                                        ]}
                                        selectedValue={dataSharing}
                                        onSelectChange={this.onSelectChange}
                                        disabled={model.getDataRowSetting() !== 0}
                                    />
                                </div>
                            </>
                        )}
                    </Modal>
                )}
            </>
        );
    }
}
