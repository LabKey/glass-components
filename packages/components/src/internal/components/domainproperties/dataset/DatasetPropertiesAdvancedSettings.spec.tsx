/*
 * Copyright (c) 2020 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import renderer from 'react-test-renderer';

import { mount } from 'enzyme';

import getDatasetDesign from '../../../../test/data/dataset-getDatasetDesign.json';
import getDatasetDesignSharedStudy from '../../../../test/data/dataset-getDatasetDesignSharedStudy.json';
import { NEW_DATASET_MODEL_WITHOUT_DATASPACE } from '../../../../test/data/constants';
import { SelectInput } from '../../../../index';

import { AdvancedSettings, DatasetSettingsInput, DatasetSettingsSelect } from './DatasetPropertiesAdvancedSettings';
import { DatasetModel } from './models';

const newDatasetModel = DatasetModel.create(NEW_DATASET_MODEL_WITHOUT_DATASPACE, undefined);
const datasetModel = DatasetModel.create(null, getDatasetDesign);
const sharedDatasetModel = DatasetModel.create(null, getDatasetDesignSharedStudy);

describe('Dataset Advanced Settings', () => {
    test('New Dataset, without dataspace options', () => {
        const datasetAdvancedSetting = (
            <AdvancedSettings title="Advanced Settings" model={newDatasetModel} applyAdvancedProperties={jest.fn()} />
        );

        const dom = renderer.create(datasetAdvancedSetting).toJSON();
        expect(dom).toMatchSnapshot();
    });

    test('New Dataset, with dataspace options', () => {
        const datasetAdvancedSetting = (
            <AdvancedSettings title="Advanced Settings" model={newDatasetModel} applyAdvancedProperties={jest.fn()} />
        );

        const dom = renderer.create(datasetAdvancedSetting).toJSON();
        expect(dom).toMatchSnapshot();
    });

    test('Edit Dataset, without dataspace options', () => {
        const datasetAdvancedSetting = (
            <AdvancedSettings title="Advanced Settings" model={datasetModel} applyAdvancedProperties={jest.fn()} />
        );

        const dom = renderer.create(datasetAdvancedSetting).toJSON();
        expect(dom).toMatchSnapshot();
    });

    test('Edit Dataset, with dataspace options', () => {
        const datasetAdvancedSetting = (
            <AdvancedSettings
                title="Advanced Settings"
                model={sharedDatasetModel}
                applyAdvancedProperties={jest.fn()}
            />
        );

        const dom = renderer.create(datasetAdvancedSetting).toJSON();
        expect(dom).toMatchSnapshot();
    });

    test('DatasetSettingsInput', () => {
        const datasetSettingsInput = mount(
            <DatasetSettingsInput
                required={true}
                name="name"
                label="Name"
                value={name}
                helpTip={<>Help tip</>}
                placeholder="Enter a name for this dataset"
                disabled={false}
                onValueChange={jest.fn()}
                showInAdvancedSettings={false}
            />
        );

        expect(datasetSettingsInput.props().label).toEqual('Name');
        expect(datasetSettingsInput.props().placeholder).toEqual('Enter a name for this dataset');
        datasetSettingsInput.unmount();
    });

    test('DatasetSettingsSelect', () => {
        const datasetSettingsSelect = mount(
            <DatasetSettingsSelect
                name="visitDateColumn"
                label="Visit Date Column"
                helpTip={<>Help tip</>}
                selectOptions={[{ label: 'A', value: 1 }]}
                selectedValue={1}
                onSelectChange={jest.fn()}
            />
        );

        expect(datasetSettingsSelect.find(SelectInput)).toHaveProperty('name');
        datasetSettingsSelect.unmount();
    });
});
