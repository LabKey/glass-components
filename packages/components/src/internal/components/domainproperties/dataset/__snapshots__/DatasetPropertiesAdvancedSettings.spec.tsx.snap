// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dataset Advanced Settings Edit Dataset, with dataspace options 1`] = `
<button
  className="domain-field-float-right btn btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Advanced Settings
</button>
`;

exports[`Dataset Advanced Settings Edit Dataset, without dataspace options 1`] = `
<button
  className="domain-field-float-right btn btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Advanced Settings
</button>
`;

exports[`Dataset Advanced Settings New Dataset, with dataspace options 1`] = `
<button
  className="domain-field-float-right btn btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Advanced Settings
</button>
`;

exports[`Dataset Advanced Settings New Dataset, without dataspace options 1`] = `
<button
  className="domain-field-float-right btn btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Advanced Settings
</button>
`;
